File: _p1_problems/Min Stack/MinStack.java
public class MinStack {
    private Stack<Integer> stack = new Stack<>();
    private Stack<Integer> minStack = new Stack<>();

    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty() || x <= minStack.peek()) {
            minStack.push(x);
        }
    }

    public void pop() {
        if (stack.isEmpty()) throw new RuntimeException();
        if (stack.pop().equals(minStack.peek())) minStack.pop();
    }

    public int top() {
        if (stack.isEmpty()) throw new RuntimeException();
        return stack.peek();
    }

    public int getMin() {
        if (minStack.isEmpty()) throw new RuntimeException();
        return minStack.peek();
    }
}

File: _designs_001/Min Stack/MinStack.rb
class MinStack
  def initialize
    @stack, @min_stack = [], []
  end

  def push(x)
    @stack.push(x)
    @min_stack.push(x) if (@min_stack.empty? or @min_stack.last >= x)
  end

  def pop
    x = @stack.pop
    @min_stack.pop if @min_stack.last == x
    x
  end

  def top
    @stack.last
  end

  def getMin
    @min_stack.last
  end
end